### ISSUE TITLE
boolean ArgumentParser arguments with default value reports wrong value

### ISSUE THREAD

**Issue by Jim-Holmstroem** [guildai/guildai#124]
```
# bool_arg.py
import argparse

parser = argparse.ArgumentParser()

parser.add_argument('--foo', default=False, type=bool)

print(parser.parse_args())
```

I know that it's not the right way to handle default values for booleans with argparse (kinda weird behavior, even without guild) and that you should use `store_false` instead, but guild reports the wrong thing `foo: no` when its real value is True, which can really mess up your debugging efforts
```
(env) $ guild run bool_arg foo=False
You are about to run m:bool_arg
  foo: no
Continue? (Y/n) y
Namespace(foo=True)
(env) $ guild run bool_arg foo=True
You are about to run m:bool_arg
  foo: yes
Continue? (Y/n) y
Namespace(foo=True)
```

```
# guild.yml
- package: p
  python-requires: ==3.6
- model: m
  sourcecode:
  - '*.py'
  - exclude:
      dir: input
  operations:
    bool_arg:
        main: bool_arg
  resources:
    input:
      sources:
        - input
```
```
guild_version:             0.7.0.rc2
guild_install_location:    /home/jim/Programming/utter-challenge/venv/lib/python3.6/site-packages/guild
guild_home:                /home/jim/Programming/utter-challenge/venv/.guild
guild_resource_cache:      /home/jim/.guild/cache/resources
installed_plugins:         cpu, disk, exec_script, gpu, keras, memory, perf, python_script, queue, skopt
python_version:            3.6.9 (default, Nov  7 2019, 10:44:02) [GCC 8.3.0]
python_exe:                /home/jim/Programming/utter-challenge/venv/bin/python3.6
platform:                  Linux 5.3.0-26-generic x86_64
tensorboard_version:       2.1.0
cuda_version:              9.1.85
nvidia_smi_version:        430.64
latest_guild_version:      0.6.6
```

potentially related to #99 but I haven't looked into it

**Comment by gar1t** [comment]
Yeah, this is a bug in Guild. Boolean `False` values to Python apparently have to be passed as empty strings. This would cause me to NEVER use boolean types as even the very reasonable args `False` and `0` are treated as `True` by argparse.

```
$ python bool_arg.py --foo False
Namespace(foo=True)
$ python bool_arg.py --foo 0
Namespace(foo=True)
```

Unless you're prohibited, I recommend using the `store_true` or `store_false` pattern, which is more typical a boolean interface. Next I'd use an `int` or `string` (default) type or a custom type (you can specify your own function) and check values explicitly. The `boolean` conversion logic from `argparse` blindly passes a string to the `boolean` function and thus this behavior from Python.

Here's how I'd do this, barring some other requirement that would prohibit it:

``` python
import argparse

parser = argparse.ArgumentParser()

parser.add_argument('--foo', default=False, type=bool)  # not this :)
parser.add_argument('--bar', action='store_true')  # this when default is True
parser.add_argument('--baz', action='store_false')  # this when default is False

print(parser.parse_args())
```

This should give you the behavior you're looking for.

```
$ guild run baz=no bar=yes
You are about to run m:bool_arg
  bar: yes
  baz: no
  foo: no
Continue? (Y/n) Namespace(bar=True, baz=False, foo=True)
```

This behavior can be explicitly defined for a flag using the `arg-switch` flag attr, which specifies the flag value that triggers the use of an option _switch_ (i.e. an option arg without a value).

``` yaml
op:
  flags:
    bar:
      type: boolean
      default: no
      arg-switch: yes
```

**Comment by Jim-Holmstroem** [comment]
yeah we went with type=int, which worked as expected
just lost some time into debugging (should have known this issue, but I'm kind relearning python again after a few years of Rust
haven't looked closely on guilds implementation of argparse handling but would it be hard to have a warning like "argument with type bool not supported"?

**Comment by gar1t** [comment]
If not sure what the right course is. Guild could pass an empty string to accommodate the `boolean` type conversion. I do think that's a broken interface (no user would know to do this in an unwrapped Python script) so it's accommodating a scenario that's not just very rare but probably shouldn't be used.

That said, Guild does support argparse and boolean is a flag type. So I think the correct change to Guild here should be to pass an empty string for False and I guess "yes" for True.

Guild could warn, but it could just as easily just accommodate the interface.

**Comment by gar1t** [comment]
Just to followup on this... we've a couple other cases come through on this and I want to make an official proclamation of strong personal opinion :)

Don't use `type=bool` when adding an `argparse` option. That creates an interface that is unusable.

Instead, use `action='store_true'` or `action='store_false'` or use a custom function ala `type=user_bool` that knows how to convert values like `'True'` and `'yes'` etc. to boolean values.

In the interest of house keeping, I'm going to close this. If there's a dispute - please anyone feel free to re-reopen.

**Comment by gar1t** [comment]
Re-opening as this is fixed in 0.7.0.rc10. Guild correctly handled argparse arg encoding for booleans, which is an empty string. While I still argue that one should not use the `bool` argument type, Guild at least will handle it correctly now.

**Comment by gar1t** [comment]
Fix is now available in 0.7.0.
